### PR TITLE
Added .env support

### PR DESCRIPTION
**Pull request by mitsuhiko**
This will instruct Flask to read .env files and then to load the
environment variables from there.  Additionally it automatically
changes the working directory to the path the .env is contained in.

### CODE REVIEW

File: flask/cli.py [hunk 1]
```diff
@@ -120,6 +120,47 @@ def find_default_import_path():
     return app
 
 
+def find_dotenv():
+    here = os.getcwd()
+    while 1:
```
Comment (ThiefMaster) [inline flask/cli.py]: `s/1/True/`

File: flask/cli.py [hunk 2]
```diff
@@ -120,6 +120,47 @@ def find_default_import_path():
...
+    if path is None:
+        return
+    os.chdir(os.path.dirname(path))
+    with open(path, 'rb') as f:
+        for line in f:
+            line = line.strip()
+            if not line or line[:1] == '#' or '=' not in line:
+                continue
+            key, value = line.split('=', 1)
+            _set_env(key, value)
```
Comment (ThiefMaster) [inline flask/cli.py]: Shouldn't existing env vars be ignored instead of being overwritten?

Comment (ThiefMaster) [inline flask/cli.py]: Would it make sense to split surrounding whitespace from both key and value?

File: flask/cli.py (hunk 1 above)
Comment (mitsuhiko) [inline flask/cli.py]: Only over my cold dead body ;)

File: flask/cli.py (hunk 2 above)
Comment (mitsuhiko) [inline flask/cli.py]: Typically `.env` files overwrite but I can see that being annoying. Will think about it.

Comment (mitsuhiko) [inline flask/cli.py]: Probably a good idea.

Comment (ThiefMaster) [inline flask/cli.py]: at least that node lib i mentioned on irc doesn't - it seems useful that you can set e.g. FLASK_DEBUG=0 on the shell even though you enable it in your .env file by default

File: flask/cli.py [hunk 3]
```diff
@@ -120,6 +120,47 @@ def find_default_import_path():
...
+def load_dotenv():
+    """Loads a dotenv file for flask."""
+    path = find_dotenv()
+    if path is None:
+        return
+    os.chdir(os.path.dirname(path))
+    with open(path, 'rb') as f:
+        for line in f:
+            line = line.strip()
+            if not line or line[:1] == '#' or '=' not in line:
```
Comment (ThiefMaster) [inline flask/cli.py]: I think this won't work in Python 3

    >>> b'#' == u'#'
    False

Comment (mitsuhiko) [inline flask/cli.py]: Indeed. Also I just noticed we should probably support quoted values.

Comment (untitaker) [inline flask/cli.py]: I think it would be a good idea to bail out on invalid lines. If somebody does this, the value will get trunchated:

```
FOO_MULTILINE="foo
bar"
```

Furthermore I'd imagine that this will break on Python 3 since you're splitting with a native string while reading in binary mode.

**Comment by mitsuhiko** [comment]
Yeah, tests will come if we like this behavior.

File: flask/cli.py (hunk 2 above)
Comment (untitaker) [inline flask/cli.py]: OTOH you can override FLASK_RUN_FROM_CLI with this.

**Comment by untitaker** [comment]
In case it gets lost: https://github.com/pallets/flask/pull/2164#discussion_r98361152

**Comment by RonnyPfannschmidt** [comment]
isnt there a library for env files? imho flask shouldnt reimplement

**Comment by mitsuhiko** [comment]
@RonnyPfannschmidt i'm not going to pull in a one liner dependency here.

@untitaker the behavior here is consistent with other libraries that implement .env loading.

File: docs/cli.rst [hunk 4]
```diff
@@ -48,6 +51,12 @@ Or with a filename::
     export FLASK_APP=/path/to/hello.py
     flask run
 
+If you use Flask 0.13 or later you can also set the environment variables
+in a ``.env`` file.  For instance like this::
+
+    FLASK_APP=/path/to/hello.py
+    FLASK_DEBUG=1
+
```
Comment (miguelgrinberg) [inline docs/cli.rst]: I have mixed feelings regarding including `FLASK_APP` in this example. The typical usage of a `.env` file is for configuration items that change between installations (i.e. local machine vs production server). However, the Flask app instance is fixed. I understand why you would want to put it here, as it annoys me to no end to have to set `FLASK_APP` to be able to use the Flask cli, but I'm thinking people will be inclined to add the `.env` file to source control only so that they can forget about `FLASK_APP`, and if they also store sensitive information such as email server passwords, AWS creds, etc. in this file (as that is one of the most interesting uses of a `.env` file IMHO) then we have a problem.

Comment (mitsuhiko) [inline docs/cli.rst]: I don't think the worry about this is any bigger than config files though, no?

Comment (ThiefMaster) [inline docs/cli.rst]: True... being able to commit a file specifying the app's location sounds quite appealing

Comment (mitsuhiko) [inline docs/cli.rst]: FWIW, the filename can be relative because flask CDs.

Comment (mitsuhiko) [inline docs/cli.rst]: Also more importantly only the `flask` command loads the .env file. So this will not work for production use (intentionally).

Comment (miguelgrinberg) [inline docs/cli.rst]: All I'm saying is that the "traditional" use of a `.env` file is to store secrets that are specific to the particular installation, and that in this usage model the `.env` file should obviously not be committed to source control. This example goes against this traditional usage, by creating an incentive for users to commit the file.

Flask config files are different in my opinion. Those can have non-sensitive values. Sensitive values in config files should import from the environment (such as the `.env` file that should not be on source control).

Comment (miguelgrinberg) [inline docs/cli.rst]: > So this will not work for production use (intentionally).

And you don't think that's odd? I can do `flask db upgrade` just fine on my dev box, but when I got to my production EC2 I have to do `FLASK_APP=app flask db upgrade`?

Comment (mitsuhiko) [inline docs/cli.rst]: @miguelgrinberg i assume people deploying wheels and not their git repos. As such the `.env` file won't be there.

Comment (untitaker) [inline docs/cli.rst]: If the `.env` file isn't checked into git (as expected), that logic shouldn't apply, no?

Comment (miguelgrinberg) [inline docs/cli.rst]: Let me summarize. It will be very tempting to use the `.env` just so that you don't have to think about `FLASK_APP` anymore. As such, one might be inclined to commit this `.env` file to git, so that you don't have to make one in every machine the app is installed. Since many other frameworks use a `.env` file as a place to store secrets, this presents a conflict, in my opinion. We would be using the .env file in a different way that what I consider the norm.

As another issue, having ` .env` file that is not imported when you run the app in production via gunicorn, etc. is also a problem, if you use it to store other settings that are needed by the application, besides `FLASK_APP` and `FLASK_DEBUG`.

For `FLASK_APP`, I would rather implement some sort of default, so that you don't have to rely on configuration. For example, I would be okay with `FLASK_APP` defaulting to `wsgi:application`. You then write a two line `wsgi.py`, safely commit it to git and be done with it.

Comment (mitsuhiko) [inline docs/cli.rst]: I'm not sure sure I understand. Why would anyone get the idea to put other things into that `.env` file since it won't ever make it to deploy? (Unless people deploy out of git repos which comes with its own set of issues).

Comment (miguelgrinberg) [inline docs/cli.rst]: Because that's how a .env file is used in a lot of other frameworks. That's the first thing I think about with `.env` files. I actually have implemented a `.env` importer for my own Flask projects as well, specifically for secrets. It's a pretty standard practice, I think.

Comment (mitsuhiko) [inline docs/cli.rst]: Where does the `.env` file live then?

Comment (ThiefMaster) [inline docs/cli.rst]: I think for the production usecase (`flask db upgrade`) we could even go as far as recommending people to use a custom command (#2163). However, while that deals with the app chances are good that a well-written application uses an environment variable to load its config file.

Not sure if that's within Flask's scope though - adding an `export MYAPP_CFG=/path/to/myapp.cfg` to the `.profile` on the production machine might be the cleaner solution for that.

**Comment by mitsuhiko** [comment]
So maybe we put the conversation back to where this started: the goal was to make it easier during development to refer to the current Flask app. I do not see this as a feature for production. However let's see if we can figure out the production case as well.

How does one refer to their own app in production without exporting env variables?

File: docs/cli.rst (hunk 4 above)
Comment (miguelgrinberg) [inline docs/cli.rst]: @mitsuhiko the `.env` file doesn't live anywhere. Some projects include a `.env.template` or similar file that you can use as a template to write your own. You are supposed to write one in each installation of your app. I've seen people write these files manually, even on production deploys. I've also seen them generated by fabric/ansible/etc.

@ThiefMaster Even with a custom command, you need to know where the app is, right? I don't see how this relates to selecting a config file via environment. All the config file stuff happens after the app exists, so nothing you put in the config file will help you locate the app.

Comment (ThiefMaster) [inline docs/cli.rst]: Installing the app would register the entry point pointing to the app / app factory, so the `myapp` command would know where the app is.

Comment (miguelgrinberg) [inline docs/cli.rst]: @ThiefMaster Ah, okay, I think I understand now. What I don't understand is why in your PR you are using this technique just for additional commands. Why can't this be an alternative to feed the app to the main Flask cli?

Comment (ThiefMaster) [inline docs/cli.rst]: Imagine two flask apps using this that are installed in the same environment - how would you know which one to use? Also, I think for bigger applications it's nice to have a custom command named like the application without losing any of the `flask` command's benefits.

**Comment by miguelgrinberg** [comment]
@mitsuhiko I agree, good idea to pull back a bit.

I think referring to the app instance should work in the same way for development and production, I don't see why the methods should differ, since the app instance is the same in all cases. And there is also the complication of app factories, which are not friendly to `FLASK_APP` (https://github.com/pallets/flask/issues/2027).

What I do, is define a `wsgi.py` file for each of my projects. I started adopting this as a convention, so I can have a hardcoded `FLASK_APP=wsgi.app` set in my `.profile` file and switch from app to app without worries. Would you consider implementing one or more sensible defaults for `FLASK_APP`? I think `wsgi.application` and `wsgi.app` and fairly common and could be defaults.

**Comment by jeffwidman** [comment]
I've also commonly seen `.env` files used to store secrets, and managed via Ansible Vault or similar.

File: CHANGES [hunk 5]
```diff
@@ -11,6 +11,9 @@ Major release, unreleased
 - Make `app.run()` into a noop if a Flask application is run from the
   development server on the command line.  This avoids some behavior that
   was confusing to debug for newcomers.
+- The `flask` command line tool now loads a `.env` file when it discovers
+  one.  When it finds one it will change into that folder before starting
```
Comment (JordanP) [inline CHANGES]: It will "change", change what ?

Comment (ThiefMaster) [inline CHANGES]: `chdir()`

**Comment by davidism** [comment]
> How does one refer to their own app in production without exporting env variables?

I do it by not using the `flask` command directly. Instead, I install an entry point to the project's own `FlaskGroup`, configured appropriately, and expect the project to be installed in the env both in development and production.

~~~python
cli = FlaskGroup(create_app=create_app)
~~~

~~~python
setup(
    ...,
    entry_points={
        'console_scripts': [
            'bwportal=bwportal.cli:cli.main',
        ],
    },
    ...,
)
~~~

**Comment by davidism** [comment]
One option for making this play nicer with the factory pattern is to allow making a call in the `FLASK_APP` var. Gunicorn and uWSGI allow this. `gunicorn bwportal.app:create_app('dev')`

**Comment by nickjj** [comment]
@mitsuhiko

> How does one refer to their own app in production without exporting env variables?

Is this a valid question? Because you could also say "how does one refer to their own app in production without creating an instance/settings.py file", and then it becomes the same thing.

At the end of the day, this file needs to be on the production box for the app to be usable.

Personally, I use Docker and it has mechanisms for supplying an `.env` file of my choosing, then it makes those env variables available to the Flask app with no extra dirty work.

Without Docker, I would likely drop an environment file in `/etc/default/fooapp` which is the defaco-standard place on a Debian system to write out app-level environment variables. Then I would source that file in whatever init system I use.

In other words, in both development and production this env file would be available to the Flask app.

If this feature does end up being implemented, please let us configure the `.env` filename.

I would much prefer to use env variables instead of config/instance files because it plays much nicer when trying to scale applications on infrastructures that let you define clusters of servers, but then can define env variables in a central area (outside of each individual host), and then they become available to all hosts without worrying about the logistics of getting an env file on the system.

This could be done now of course, but it kind of goes against the norm which is to use config/instance based config files.

**Comment by justanr** [comment]
If I'm understanding this correctly, this would only read the `.env` when executed from the CLI, correct? That seems like it setting developers up for mistakes, I can see the bug reports and SO questions now: "Flask loads my .env file when I run the development server, but not when executed from uWSGI"

If this was going to be added, I'd rather see it implemented into either the Flask class or config class to at least be consistent between server options.

File: docs/cli.rst (hunk 4 above)
Comment (greyli) [inline docs/cli.rst]: @ThiefMaster Adding this into virtual environment's activate script will solve this problem.

**Comment by greyli** [comment]
I did it by adding this into virtual environment's `activate script`, the variable will be set automatically when the virtual environment was activated. No multiple application crashing, no new file.

For example:  
Unix Bash: `venv/bin/activate`:
```
FLASK_APP=foo.py
export FLASK_APP
```
Windows CMD.exe: `venv\Scripts\activate.bat`:
```
set "FLASK_APP=foo.py" 
:END
```

**Comment by untitaker** [comment]
I don't recommend editing `activate` scripts.

**Comment by greyli** [comment]
@untitaker More detail?

**Comment by untitaker** [comment]
I already elaborated in the PR you sent. You are editing random files which will get overwritten if you recreate the virtualenv (e.g. because it grew too big, or because a Python upgrade trashed it).

**Comment by davidism** [comment]
How about adding an entry point in setup.py, similar to how we can add extra commands?

**Comment by untitaker** [comment]
See https://github.com/pallets/flask/pull/2234#issuecomment-292532462

**Comment by greyli** [comment]
#2234

**Comment by davidism** [comment]
I'm leaning on the side of rejecting this. I agree that it would cause confusion, and we already have a documented way to set the env when working locally.

**Comment by untitaker** [comment]
Hmm, I'm not sure if we have discussed this enough yet. I only merged #2234
because it enhanced the docs for existing recommendations.

I would actually prefer a file separate from the virtualenv activate script,
for the additional reason that you would have to edit it for multiple shells.
